src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs

```diff
@@ -0,0 +1,1277 @@
+
+        void IHttpHeadersHandler.OnHeader(ReadOnlySpan<byte> name, ReadOnlySpan<byte> value)
+        {
+            Debug.Assert(name.Length > 0);
+            if (!HeaderDescriptor.TryGet(name, out HeaderDescriptor descriptor))
+            {
+                // Invalid header name
+                throw new HttpRequestException(SR.Format(SR.net_http_invalid_response_header_name, Encoding.ASCII.GetString(name)));
+            }
+            OnHeader(staticIndex: null, descriptor, staticValue: default, literalValue: value);
+        }
+
+        void IHttpHeadersHandler.OnStaticIndexedHeader(int index)
+        {
+            GetStaticQPackHeader(index, out HeaderDescriptor descriptor, out string knownValue);
+            OnHeader(index, descriptor, knownValue, literalValue: default);
+        }
+
+        void IHttpHeadersHandler.OnStaticIndexedHeader(int index, ReadOnlySpan<byte> value)
+        {
+            GetStaticQPackHeader(index, out HeaderDescriptor descriptor, knownValue: out _);
+            OnHeader(index, descriptor, staticValue: null, literalValue: value);
+        }
+
+        private void GetStaticQPackHeader(int index, out HeaderDescriptor descriptor, out string knownValue)
+        {
+            if (!HeaderDescriptor.TryGetStaticQPackHeader(index, out descriptor, out knownValue))
+            {
+                if (NetEventSource.IsEnabled) Trace($"Response contains invalid static header index '{index}'.");
+                throw new Http3ConnectionException(Http3ErrorCode.ProtocolError);
+            }
+        }
+
+        /// <param name="staticIndex">The static index of the header, if any.</param>
+        /// <param name="descriptor">A descriptor for either a known header or unknown header.</param>
+        /// <param name="staticValue">The static indexed value, if any.</param>
+        /// <param name="literalValue">The literal ASCII value, if any.</param>
+        /// <remarks>One of <paramref name="staticValue"/> or <paramref name="literalValue"/> will be set.</remarks>
+        private void OnHeader(int? staticIndex, HeaderDescriptor descriptor, string staticValue, ReadOnlySpan<byte> literalValue)
+        {
+            if (descriptor.Name[0] == ':')
+            {
+                if (descriptor.KnownHeader != KnownHeaders.PseudoStatus)
+                {
+                    if (NetEventSource.IsEnabled) Trace($"Received unknown pseudo-header '{descriptor.Name}'.");
+                    throw new Http3ConnectionException(Http3ErrorCode.ProtocolError);
+                }
+
+                if (_headerState != HeaderState.StatusHeader)
+                {
+                    if (NetEventSource.IsEnabled) Trace("Received extra status header.");
+                    throw new Http3ConnectionException(Http3ErrorCode.ProtocolError);
+                }
+
+                int statusCode = staticIndex switch
```

Comment:
I thought we had a table for this in H3StaticTable?

Reply:
I believe we had the reverse table, but I could be wrong.